hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -137,7 +163,47 @@ protected String tableName() {
 
   @Override
   public FileIO io() {
-    return fileIO;
+    if (encryptionManager == null) {
+      encryptionManager = encryption();
+    }
+
+    if (!encryptedTable) {
+      return fileIO;
+    }
+
+    if (encryptingFileIO != null) {
+      return encryptingFileIO;
+    }
+
+    encryptingFileIO = EncryptingFileIO.combine(fileIO, encryptionManager);
+    return encryptingFileIO;
+  }
+
+  @Override
+  public EncryptionManager encryption() {
+    if (encryptionManager != null) {
+      return encryptionManager;
+    }
+
+    if (encryptionKeyId == null) {
+      encryptionPropsFromHms();
```

Comment:
This LGTM now.

The previous approach (i.e. call the `dekLength()` in `BaseMetastoreTableOperations`) will only set `encryptionDekLength` on the write path (the `writeNewMetadataFile` method). Therefore, Hive Catalog readers get the default -1 value.